### PR TITLE
Add enzyme_width attribute

### PR DESCRIPTION
**Pull request by tgymnich**
- Added enzyme_width attribute
- Pass width through enzyme logic

### CODE REVIEW

File: enzyme/Enzyme/Enzyme.cpp [hunk 1]
```diff
@@ -357,12 +434,48 @@ class Enzyme : public ModulePass {
     auto FT = cast<Function>(fn)->getFunctionType();
     assert(fn);
 
-    bool sret = CI->paramHasAttr(0, Attribute::StructRet) ||
-                cast<Function>(fn)->hasParamAttribute(0, Attribute::StructRet);
-
     IRBuilder<> Builder(CI);
-
+    std::optional<unsigned int> width;
```
Comment (wsmoses) [inline enzyme/Enzyme/Enzyme.cpp:438]: Probably better to use the llvm::Optional

File: enzyme/Enzyme/Enzyme.cpp [hunk 2]
```diff
@@ -357,12 +434,48 @@ class Enzyme : public ModulePass {
...
+      if (getMetadataName(arg) == "enzyme_width") {
+        assert(mode == DerivativeMode::ForwardMode);
+
+        if (width) {
+          EmitFailure("IllegalVectorWidth", CI->getDebugLoc(), CI,
+                      "vector width declared more than once",
+                      *CI->getArgOperand(i), " in", *CI);
+          return false;
+        }
+
```
Comment (wsmoses) [inline enzyme/Enzyme/Enzyme.cpp:454]: Do a bounds check here beforehand?

File: enzyme/Enzyme/Enzyme.cpp [hunk 3]
```diff
@@ -357,12 +434,48 @@ class Enzyme : public ModulePass {
...
+        Value *width_arg = CI->getArgOperand(i + 1);
+        if (auto cint = dyn_cast<ConstantInt>(width_arg)) {
+          width = cint->getSExtValue();
```
Comment (wsmoses) [inline enzyme/Enzyme/Enzyme.cpp:457]: ZExt since negative value doesn't make sense?

File: enzyme/Enzyme/Enzyme.cpp [hunk 4]
```diff
@@ -447,113 +589,61 @@ class Enzyme : public ModulePass {
       }
       assert(truei < FT->getNumParams());
       auto PTy = FT->getParamType(truei);
-      DIFFE_TYPE ty = DIFFE_TYPE::CONSTANT;
+      DIFFE_TYPE ty = whatType(PTy, mode);
+      std::optional<StringRef> metaString = getMetadataName(res);
```
Comment (wsmoses) [inline enzyme/Enzyme/Enzyme.cpp:593]: llvm::Optional

File: enzyme/Enzyme/Enzyme.cpp (hunk 1 above)
Comment (wsmoses) [reply enzyme/Enzyme/Enzyme.cpp:438]: Or perhaps even better have it be a regular integer, default set to 0 as uninitialized.

File: enzyme/Enzyme/Enzyme.cpp [hunk 5]
```diff
@@ -1427,7 +1548,7 @@ class Enzyme : public ModulePass {
                        Arch == Triple::amdgcn;
 
       auto val = GradientUtils::GetOrCreateShadowConstant(
-          Logic, TLI, TA, fn, pair.second, AtomicAdd, PostOpt);
+          Logic, TLI, TA, fn, pair.second, 1, AtomicAdd, PostOpt);
```
Comment (wsmoses) [inline enzyme/Enzyme/Enzyme.cpp:1551]: should this not be width?

File: enzyme/Enzyme/EnzymeLogic.cpp [hunk 6]
```diff
@@ -3190,8 +3190,8 @@ Function *EnzymeLogic::CreatePrimalAndGradient(
   bool diffeReturnArg = key.retType == DIFFE_TYPE::OUT_DIFF;
 
   DiffeGradientUtils *gutils = DiffeGradientUtils::CreateFromClone(
-      *this, key.mode, key.todiff, TLI, TA, key.retType, diffeReturnArg,
-      key.constant_args, retVal, key.additionalType, omp);
+      *this, key.mode, /* width */ 1, key.todiff, TLI, TA, key.retType,
```
Comment (wsmoses) [inline enzyme/Enzyme/EnzymeLogic.cpp:3193]: This should also have the width passed through.

File: enzyme/Enzyme/Enzyme.cpp (hunk 5 above)
Comment (tgymnich) [reply enzyme/Enzyme/Enzyme.cpp:1551]: This is still an open question. On the one hand a function should only be shadowed once, on the other hand fwd mode currently shadows it n times and only uses the first. Also we don't necessarily know the width at this point. We would need to add width metadata to __enzyme_virtualreverse somehow. Lets keep that for later.

File: enzyme/Enzyme/Enzyme.cpp (hunk 3 above)
Comment (tgymnich) [reply enzyme/Enzyme/Enzyme.cpp:457]: fixed

File: enzyme/Enzyme/Enzyme.cpp (hunk 1 above)
Comment (tgymnich) [reply enzyme/Enzyme/Enzyme.cpp]: done

File: enzyme/Enzyme/EnzymeLogic.cpp (hunk 6 above)
Comment (tgymnich) [reply enzyme/Enzyme/EnzymeLogic.cpp]: done

File: enzyme/Enzyme/Enzyme.cpp (hunk 4 above)
Comment (tgymnich) [reply enzyme/Enzyme/Enzyme.cpp:593]: replaced

File: enzyme/Enzyme/Enzyme.cpp (hunk 2 above)
Comment (tgymnich) [reply enzyme/Enzyme/Enzyme.cpp:455]: good catch

**Review by wsmoses — APPROVED** [review]

File: enzyme/Enzyme/CApi.cpp [hunk 7]
```diff
@@ -391,17 +391,18 @@ LLVMValueRef EnzymeCreateForwardDiff(
...
       uncacheable_args, PostOpt));
 }
 LLVMValueRef EnzymeCreatePrimalAndGradient(
     EnzymeLogicRef Logic, LLVMValueRef todiff, CDIFFE_TYPE retType,
     CDIFFE_TYPE *constant_args, size_t constant_args_size,
     EnzymeTypeAnalysisRef TA, uint8_t returnValue, uint8_t dretUsed,
-    CDerivativeMode mode, LLVMTypeRef additionalArg, CFnTypeInfo typeInfo,
-    uint8_t *_uncacheable_args, size_t uncacheable_args_size,
-    EnzymeAugmentedReturnPtr augmented, uint8_t AtomicAdd, uint8_t PostOpt) {
+    CDerivativeMode mode, unsigned width, LLVMTypeRef additionalArg,
```
Comment (vchuravy) [inline enzyme/Enzyme/CApi.cpp:402]: Please open associate PR to Enzyme.jl on such API changes.,

Comment (tgymnich) [reply enzyme/Enzyme/CApi.cpp:402]: https://github.com/wsmoses/Enzyme.jl/pull/179
https://github.com/rust-ml/oxide-enzyme/pull/12

**Pull request merged**